webrender/src/util.rs

```diff
@@ -508,3 +508,67 @@ pub fn world_rect_to_device_pixels(
     let device_rect = rect * device_pixel_scale;
     device_rect.round_out()
 }
+
+/// Run a first callback over all elements in the array. If the callback returns true,
+/// The element is removed from the array and moved to a second callback.
+///
+/// This is a simple implementation waiting for Vec::drain_filter to be stable.
+/// When that happens, code like:
+///
+/// ```
+/// let filter = |&mut op| {
+///     match *op {
+///         Foo | Bar => true,
+///         Baz => false,
+///     }
+/// };
+/// drain_filter(
+///     &mut ops,
+///     filter,
+///     |op| {
+///         match op {
+///             Foo => { foo(); }
+///             Bar => { bar(); }
+///             Baz => { unreachable!(); }
+///         }
+///     },
+/// );
+/// ```
+///
+/// Can be rewritten as:
+///
+/// ```
+/// let filter = |&mut op| {
+///     match *op {
+///         Foo | Bar => true,
+///         Baz => false,
+///     }
+/// };
+/// for op in ops.drain_filter(filter) {
+///     match op {
+///         Foo => { foo(); }
+///         Bar => { bar(); }
+///         Baz => { unreachable!(); }
+///     }
+/// }
+/// ```
+///
+/// See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter
+pub fn drain_filter<T, F1, F2>(
+    vec: &mut Vec<T>,
+    mut filter: F1,
+    mut action: F2,
+)
+where
+    F1: FnMut(&mut T) -> bool,
+    F2: FnMut(T)
+{
+    let mut i = 0;
+    while i != vec.len() {
+        if filter(&mut vec[i]) {
+            action(vec.remove(i));
```

Comment:
This is quadratic asymptotic complexity, or rather O(N*M) where M is the number of filter matches. Can we do better?

Reply:
We can but should we? We'll get the the optimized version for free when the std implementation stabilizes and I am certain this will not show up in profiles.
To be honest it also rubs me the wrong way and I'd probably have made more of an effort if we weren't getting the better implementation through std eventually.

Reply:
ok, that's fine by me, assuming there is a limited array size on the input

Reply:
Right now the size of resource updates vector is 0 the vast majority of times, 3 while playing a video. occasionally 1 when loading an image, etc.